### DISCUSSION TITLE
How do I get hold of NuGet.VisualStudio.Contracts.INuGetProjectService ?

### DISCUSSION THREAD

**Discussion by ekrogh** [VsixCommunity/Community.VisualStudio.Toolkit#153] [Q&A]
Visual Studio (2022) claims that:

> <X> [deprecated] interface NuGetVisualStudio.IVsPackagelnstallerServices
> Contains methods to query for installed packages within the current solution.
> CS061& 'IVsPackagelnstallerServices' is obsolete: 'Use INuGetProjectService in the NuGet.VisualStudio.Contracts package instead.'
> Show potential fixes (Alt*Enter or Ctrl*.)

But if I try 
```
var tst =	await VS.GetMefServiceAsync<NuGet.VisualStudio.Contracts.INuGetProjectService>();
```

it thwows this:

> Microsoft.VisualStudio.Composition.CompositionFailedException
>   HResult=0x80131500
>   Message=Expected 1 export(s) with contract name "NuGet.VisualStudio.Contracts.INuGetProjectService" but found 0 after applying applicable constraints.
>   Source=Microsoft.VisualStudio.Composition
>   StackTrace:
>    at Microsoft.VisualStudio.Composition.ExportProvider.GetExports(ImportDefinition importDefinition)
>    at Microsoft.VisualStudio.Composition.ExportProvider.GetExports[T,TMetadataView](String contractName, ImportCardinality cardinality)
>    at Microsoft.VisualStudio.Composition.ExportProvider.GetExport[T,TMetadataView](String contractName)
>    at Microsoft.VisualStudio.Composition.ExportProvider.GetExportedValue[T]()
>    at Microsoft.VisualStudio.ComponentModelHost.ComponentModel.GetService[T]()
>    at Community.VisualStudio.Toolkit.VS.<GetMefServiceAsync>d__42`1.MoveNext()
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at xxx.CheckSolutionType.<ThisIsXamarinAsync>d__0.MoveNext() in D:\Users\eigil\projects\VSIX\xxx\xxx\Helpers\CheckSolutionType.cs:line 24
> 
>   This exception was originally thrown at this call stack:
>     [External Code]
>     xxx.CheckSolutionType.ThisIsXamarinAsync() in CheckSolutionType.cs

How do I get hold of NuGet.VisualStudio.Contracts.INuGetProjectService ?

**Comment by reduckted** [comment]
Here's how it's used in the Roslyn project: 
https://github.com/dotnet/roslyn/blob/e079a36f3ade7cda2a464e4fdcde0a11577fd1ea/src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs#L533-L540

Here's an example:
```cs
IBrokeredServiceContainer serviceContainer;
IServiceBroker serviceBroker;
INuGetProjectService nugetService;

serviceContainer = await VS.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
serviceBroker = serviceContainer?.GetFullAccessServiceBroker();
nugetService = await serviceBroker.GetProxyAsync<INuGetProjectService>(NuGetServices.NuGetProjectServiceV1, cancellationToken: default);

using (nugetService as IDisposable)
{
    InstalledPackagesResult result;

    result = await nugetService.GetInstalledPackagesAsync(projectId, CancellationToken.None);

    foreach (NuGetInstalledPackage package in result.Packages)
    {
        Debug.WriteLine(package.Id);
    }
}
```

**Comment by workgroupengineering** [comment]
Do you know if there is a way to get notified when one or more nuget packages are installed / removed without polling?
